src/consumer/worker.js

```diff
@@ -0,0 +1,34 @@
+/**
+ * @typedef {(batch: T, metadata: { workerId: number }) => Promise<void>} Handler
+ * @template T
+ *
+ * @typedef {ReturnType<typeof createWorker>} Worker
+ */
+
+/**
+ * @param {{ handler: Handler<T>, workerId: number }} options
+ * @template T
+ */
+const createWorker = ({ handler, workerId }) => {
+  const getWorkerId = () => workerId
+
+  /**
+   * Takes messages from next() until it returns undefined.
+   *
+   * @param {{ next: () => T | undefined }} param0
+   * @returns {Promise<void>}
+   */
+  const run = async ({ next }) => {
+    const batch = next()
+    if (!batch) {
+      return
+    }
+
+    await handler(batch, { workerId })
+    return run({ next })
```

Comment:
If I understand this correctly, the worker is eating off of the workerQueue, and the workerQueue will contain just the batches from this fetch iteration and then resolve once the queue is empty, after which the fetcher will fetch the next round of batches, making this not infinite (which is a good thing, or I'd be concerned about memory leaks).